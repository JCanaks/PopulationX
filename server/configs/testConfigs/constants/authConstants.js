export const SIGNUP_USER = `
    mutation signup($name: String!, $email: String!, $password: String!) {
      signup(name: $name, email: $email, password: $password) {
        user {
          name
          email
        }
      }
    }
`;

export const LOGIN_USER = `
    mutation login($email: String!, $password: String!) {
      login(email: $email, password: $password) {
        user {
          name
          email
        }
      }
    }
`;

export const GET_WELCOME_MESSAGE = `
query {
  info
}`;
