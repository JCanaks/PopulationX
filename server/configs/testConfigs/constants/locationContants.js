export const CREATE_LOCATION = `
    mutation createLocation($name: String!, $noOfFemales: Int!, $noOfMales: Int!, $locatedIn: String) {
      createLocation(name: $name, noOfFemales: $noOfFemales, noOfMales: $noOfMales, locatedIn: $locatedIn) {
        name
        noOfFemales
        noOfMales
        totalResidents
        parentLocation {
          name
        }
        childLocations {
          name
        }
      }
    }
`;

export const UPDATE_LOCATION = `
    mutation updateLocation($id: ID!, $name: String, $noOfFemales: Int, $noOfMales: Int, $locatedIn: String) {
      updateLocation(id: $id, name: $name, noOfFemales: $noOfFemales, noOfMales: $noOfMales, locatedIn: $locatedIn) {
        name
        noOfFemales
        noOfMales
        totalResidents
        parentLocation {
          name
        }
        childLocations {
          name
        }
      }
    }
`;

export const GET_LOCATIONS = `
    query {
        allLocations {
        name
        noOfFemales
        noOfMales
        totalResidents
        parentLocation {
          name
        }
        childLocations {
          name
        }
      }
    }
`;

export const DELETE_LOCATION = `
    mutation deleteLocation($id: ID!) {
        deleteLocation(id: $id) {
        name
      }
    }
`;
