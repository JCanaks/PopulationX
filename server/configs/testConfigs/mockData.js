export default {
  signupInput: {
    validUserInput: {
      name: 'TestUser Signup',
      email: 'jsignup@mail.com',
      password: '1111',
    },
    invalidUserInput: {
      name: '',
      email: '',
      password: '',
    },
  },
  loginInput: {
    validUserInput: {
      email: 'jsignup@mail.com',
      password: '1111',
    },
    invalidUserInput: {
      email: '',
      password: '',
    },
  },
};
