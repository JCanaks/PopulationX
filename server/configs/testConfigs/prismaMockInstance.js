import { makePrismaClientClass } from 'prisma-client-lib';
import { typeDefs } from '../../generated/prisma-client/prisma-schema';
import { models } from '../../generated/prisma-client';

const PrismaMockInstance = makePrismaClientClass({
  typeDefs,
  models,
  endpoint: 'http://localhost:4499',
});
export default new PrismaMockInstance();
