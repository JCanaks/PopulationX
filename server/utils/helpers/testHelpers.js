import bcrypt from 'bcryptjs';

export const clearData = async (prismaMockInstance) => {
  const { deleteManyLocations, deleteManyUsers } = prismaMockInstance;
  await deleteManyLocations({});
  await deleteManyUsers({});
};

export const createTestUser = async (prismaMockInstance, signupInput) => {
  const { createUser } = prismaMockInstance;
  const password = await bcrypt.hash(signupInput.password, 10);
  const user = await createUser({ ...signupInput, password });

  return user;
};

export const getLocationId = async (prismaMockInstance, name) => {
  const location = await prismaMockInstance.location({
    name,
  });
  return location.id;
};
