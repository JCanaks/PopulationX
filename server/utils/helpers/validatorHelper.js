import Validator from 'validatorjs';

const validationErrors = {
  name: {
    required: 'Please enter a name in the specified field',
    min: 'Name should be a minimum of 2 characters',
    max: 'Name should have a maximum of 50 characters',
  },
  email: {
    required: 'Please enter an email in the specified field',
    email: 'Email entered is invalid',
  },
  password: {
    required: 'Please enter a password in the specified field',
    min: 'Password must be at least 4 characters',
    max: 'Password should have a maximum of 50 characters',
  },
  id: {
    required: 'Please enter an id in the specified field',
  },
  locatedIn: {
    min: 'LocatedIn should have a minimum of 2 characters',
    max: 'LocatedIn should have a maximum of 50 characters',
  },
};

const signupRules = (userInput) => {
  const rules = {
    name: ['required', 'min:2', 'max:50'],
    email: 'required|email',
    password: ['required', 'min:4', 'max:50'],
  };

  const validation = new Validator(userInput, rules, {
    'required.name': validationErrors.name.required,
    'required.email': validationErrors.email.required,
    'required.password': validationErrors.password.required,
    'min.name': validationErrors.name.min,
    'min.password': validationErrors.password.min,
    'max.name': validationErrors.name.max,
    'max.password': validationErrors.password.max,
    'email.email': validationErrors.email.email,
  });

  if (validation.fails()) {
    return validation.errors.all();
  }
  return false;
};

const loginRules = (userInput) => {
  const rules = {
    email: 'required|email',
    password: ['required', 'min:4', 'max:50'],
  };

  const validation = new Validator(userInput, rules, {
    'required.email': validationErrors.email.required,
    'required.password': validationErrors.password.required,
    'min.password': validationErrors.password.min,
    'max.password': validationErrors.password.max,
    'email.email': validationErrors.email.email,
  });

  if (validation.fails()) {
    return validation.errors.all();
  }
  return false;
};

const createLocationRules = (userInput) => {
  const rules = {
    name: ['required', 'min:2', 'max:50'],
    locatedIn: ['min:2', 'max:50'],
  };

  const validation = new Validator(userInput, rules, {
    'required.name': validationErrors.name.required,
    'min.name': validationErrors.name.min,
    'max.name': validationErrors.name.max,
    'min.locatedIn': validationErrors.locatedIn.min,
    'max.locatedIn': validationErrors.locatedIn.max,
  });

  if (validation.fails()) {
    return validation.errors.all();
  }
  return false;
};

const updateLocationRules = (userInput) => {
  const rules = {
    id: 'required',
    name: ['min:2', 'max:50'],
    locatedIn: ['min:2', 'max:50'],
  };

  const validation = new Validator(userInput, rules, {
    'required.id': validationErrors.id.required,
    'min.name': validationErrors.name.min,
    'max.name': validationErrors.name.max,
    'min.locatedIn': validationErrors.locatedIn.min,
    'max.locatedIn': validationErrors.locatedIn.max,
  });

  if (validation.fails()) {
    return validation.errors.all();
  }
  return false;
};

const deleteLocationRules = (userInput) => {
  const rules = {
    id: 'required',
  };

  const validation = new Validator(userInput, rules, {
    'required.id': validationErrors.id.required,
  });

  if (validation.fails()) {
    return validation.errors.all();
  }
  return false;
};

export default {
  signupRules,
  loginRules,
  createLocationRules,
  updateLocationRules,
  deleteLocationRules,
};
