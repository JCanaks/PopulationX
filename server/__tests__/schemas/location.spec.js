import '@babel/polyfill';
import prismaMockInstance from '../../configs/testConfigs/prismaMockInstance';
import mockData from '../../configs/testConfigs/mockData';
import { clearData, createTestUser, getLocationId } from '../../utils/helpers/testHelpers';
import createGraphqlTestServer from '../../configs/testConfigs/testClient';
import generateToken from '../../utils/services/tokenService';
import { CREATE_LOCATION, UPDATE_LOCATION, GET_LOCATIONS, DELETE_LOCATION } from '../../configs/testConfigs/constants/locationContants';

const { signupInput, locationInput } = mockData;
let user;
let location;

describe('Testing location resolvers', () => {
    beforeAll(async () => {
        await clearData(prismaMockInstance);
        user = await createTestUser(prismaMockInstance, signupInput.validUserInput);
    });

    afterAll(async () => {
        await clearData(prismaMockInstance);
    });
    
    test('Creating a location with valid user input', async () => {
        const request = {
            headers: {
                authorization: generateToken({ userId: user.id }, '3d'),
            }
        }
        const { mutate } = createGraphqlTestServer(request);
        const args = locationInput.validCreateLocationInput;

        const { errors, data } = await mutate({
            mutation: CREATE_LOCATION,
            variables: args,
        });
        location = data.createLocation;

        expect(data).toBeDefined();
        expect(data).toMatchSnapshot();
        expect(errors).toBeUndefined();
    });

    test('Creating a location with invalid input', async () => {
        const request = {
            headers: {
                authorization: generateToken({ userId: user.id }, '3d'),
            }
        }
    
      const { mutate } = createGraphqlTestServer(request);
      const args = locationInput.invalidCreateLocationInput;
      const { errors, data } = await mutate({
        mutation: CREATE_LOCATION,
        variables: args,
      });
      expect(data).toBeUndefined();
      expect(errors).toBeDefined();
      expect(errors).toMatchSnapshot();
    });

    test('Creating a location with name and locatedIn of invalid max value', async () => {
        const request = {
            headers: {
                authorization: generateToken({ userId: user.id }, '3d'),
            }
        }
    
      const { mutate } = createGraphqlTestServer(request);
      const args = locationInput.invalidCreateLocationInput1;
      const { errors, data } = await mutate({
        mutation: CREATE_LOCATION,
        variables: args,
      });
      expect(data).toBeUndefined();
      expect(errors).toBeDefined();
      expect(errors).toMatchSnapshot();
    });

    test('Creating a location with name and locatedIn of invalid min value', async () => {
        const request = {
            headers: {
                authorization: generateToken({ userId: user.id }, '3d'),
            }
        }
    
      const { mutate } = createGraphqlTestServer(request);
      const args = locationInput.invalidCreateLocationInput2;
      const { errors, data } = await mutate({
        mutation: CREATE_LOCATION,
        variables: args,
      });
      expect(data).toBeUndefined();
      expect(errors).toBeDefined();
      expect(errors).toMatchSnapshot();
    });

    test('Updating a location with valid user input', async () => {
        const request = {
            headers: {
                authorization: generateToken({ userId: user.id }, '3d'),
            }
        }

        const { mutate } = createGraphqlTestServer(request);
        const args = locationInput.validUpdateLocationInput;

        const id = await getLocationId(prismaMockInstance, location.name)
        const { errors, data } = await mutate({
            mutation: UPDATE_LOCATION,
            variables: {id, ...args}
        });

        location = data.updateLocation;

        expect(data).toBeDefined();
        expect(data).toMatchSnapshot();
        expect(errors).toBeUndefined();
    });

    test('Updating a location with no user input', async () => {
        const request = {
            headers: {
                authorization: generateToken({ userId: user.id }, '3d'),
            }
        }
        
        const { mutate } = createGraphqlTestServer(request);
        const { errors, data } = await mutate({
            mutation: UPDATE_LOCATION,
            variables: {}
        });
        
        expect(data).toBeUndefined();
        expect(errors).toBeDefined();
        expect(errors).toMatchSnapshot();
    });

    test('Updating a location with name and locatedIn of invalid min values', async () => {
        const request = {
            headers: {
                authorization: generateToken({ userId: user.id }, '3d'),
            }
        };

        const id = await getLocationId(prismaMockInstance, location.name)
        const { mutate } = createGraphqlTestServer(request);
        const args = locationInput.invalidUpdateLocationInput;
        const { errors } = await mutate({
            mutation: UPDATE_LOCATION,
            variables: {id, ...args}
        });
        
        expect(errors).toBeDefined();
        expect(errors).toMatchSnapshot();
    });

    test('Updating a location with name and locatedIn of invalid max values', async () => {
        const request = {
            headers: {
                authorization: generateToken({ userId: user.id }, '3d'),
            }
        };

        const id = await getLocationId(prismaMockInstance, location.name)
        const { mutate } = createGraphqlTestServer(request);
        const args = locationInput.invalidUpdateLocationInput1;
        const { errors } = await mutate({
            mutation: UPDATE_LOCATION,
            variables: {id, ...args}
        });
        
        expect(errors).toBeDefined();
        expect(errors).toMatchSnapshot();
    });

    test('Getting the list of locations available', async () => {
        const { query } = createGraphqlTestServer();
        const { errors, data } = await query({
          query: GET_LOCATIONS,
          variables: {},
        });
    
        expect(data).toBeDefined();
        expect(data).toMatchSnapshot();
        expect(errors).toBeUndefined();
      });


    test('Deleting a Location', async () => {
        const request = {
            headers: {
                authorization: generateToken({ userId: user.id }, '3d'),
            }
        }

        const { mutate } = createGraphqlTestServer(request);

        const id = await getLocationId(prismaMockInstance, location.name)
        const { errors, data } = await mutate({
            mutation: DELETE_LOCATION,
            variables: { id },
        });

        expect(data).toBeDefined();
        expect(data).toMatchSnapshot();
        expect(errors).toBeUndefined();
    });

    test('Deleting a Location with no id specified', async () => {
        const request = {
            headers: {
                authorization: generateToken({ userId: user.id }, '3d'),
            }
        }

        const { mutate } = createGraphqlTestServer(request);

        const id = '';
        const { errors, data } = await mutate({
            mutation: DELETE_LOCATION,
            variables: {id},
        });

        expect(errors).toBeDefined();
        expect(errors).toMatchSnapshot();
    });
});

