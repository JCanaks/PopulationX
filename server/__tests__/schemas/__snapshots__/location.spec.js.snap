// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing location resolvers Creating a location with invalid input 1`] = `
Array [
  [GraphQLError: Variable "$noOfFemales" of required type "Int!" was not provided.],
  [GraphQLError: Variable "$noOfMales" of required type "Int!" was not provided.],
]
`;

exports[`Testing location resolvers Creating a location with name and locatedIn of invalid max value 1`] = `
Array [
  [GraphQLError: Variable "$noOfFemales" of required type "Int!" was not provided.],
  [GraphQLError: Variable "$noOfMales" of required type "Int!" was not provided.],
]
`;

exports[`Testing location resolvers Creating a location with name and locatedIn of invalid min value 1`] = `
Array [
  [GraphQLError: Variable "$noOfFemales" of required type "Int!" was not provided.],
  [GraphQLError: Variable "$noOfMales" of required type "Int!" was not provided.],
]
`;

exports[`Testing location resolvers Creating a location with valid user input 1`] = `
Object {
  "createLocation": Object {
    "childLocations": Array [],
    "name": "lagos",
    "noOfFemales": 1000,
    "noOfMales": 950,
    "parentLocation": null,
    "totalResidents": 1950,
  },
}
`;

exports[`Testing location resolvers Deleting a Location 1`] = `
Object {
  "deleteLocation": Object {
    "name": "Lasgidi",
  },
}
`;

exports[`Testing location resolvers Deleting a Location with no id specified 1`] = `
Array [
  [GraphQLError: {"id":["Please enter an id in the specified field"]}],
]
`;

exports[`Testing location resolvers Getting the list of locations available 1`] = `
Object {
  "allLocations": Array [
    Object {
      "childLocations": Array [],
      "name": "Lasgidi",
      "noOfFemales": 2000,
      "noOfMales": 1950,
      "parentLocation": null,
      "totalResidents": 3950,
    },
  ],
}
`;

exports[`Testing location resolvers Updating a location with name and locatedIn of invalid max values 1`] = `
Array [
  [GraphQLError: {"name":["Name should have a maximum of 50 characters"],"locatedIn":["LocatedIn should have a maximum of 50 characters"]}],
]
`;

exports[`Testing location resolvers Updating a location with name and locatedIn of invalid min values 1`] = `
Array [
  [GraphQLError: {"name":["Name should be a minimum of 2 characters"],"locatedIn":["LocatedIn should have a minimum of 2 characters"]}],
]
`;

exports[`Testing location resolvers Updating a location with no user input 1`] = `
Array [
  [GraphQLError: Variable "$id" of required type "ID!" was not provided.],
]
`;

exports[`Testing location resolvers Updating a location with valid user input 1`] = `
Object {
  "updateLocation": Object {
    "childLocations": Array [],
    "name": "Lasgidi",
    "noOfFemales": 2000,
    "noOfMales": 1950,
    "parentLocation": null,
    "totalResidents": 3950,
  },
}
`;
